moda-contracts/src/Marketplace.sol

```diff
@@ -0,0 +1,236 @@
+    error SaleHasEnded(uint256 endTime, uint256 currentTime);
+    error ReleasesIsNotRegistered();
+
+    /**
+     * @dev Constructor
+     * @param treasury_ - The address of the organizations treasury
+     * @param treasuryFee_ - The percentage that will be transferred
+     * to the treasury on each sale. Based on a denominator of 10_000 e.g. 1000 = 10%
+     * @param usdc - The USDC token address
+     * @param catalog - The address of the catalog contract
+     */
+    constructor(address payable treasury_, uint256 treasuryFee_, address usdc, address catalog) {
+        if (usdc == address(0)) revert CannotBeZeroAddress();
+        if (treasuryFee_ == 0) revert TreasuryFeeCannotBeZero();
+        if (treasury_ == address(0)) revert CannotBeZeroAddress();
+        treasury = treasury_;
+        treasuryFee = treasuryFee_;
+        _usdc = usdc;
+        _catalog = catalog;
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // External Functions
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function createSale(
+        address payable beneficiary,
+        address releases,
+        uint256 tokenId,
+        uint256 amountTotal,
+        uint256 pricePerToken,
+        uint256 startAt,
+        uint256 endAt,
+        uint256 maxCountPerWallet
+    ) external {
+        if (IReleasesRegistration(_catalog).getReleasesOwner(releases) == address(0)) {
+            revert ReleasesIsNotRegistered();
+        }
+        if (beneficiary == address(0)) revert CannotBeZeroAddress();
+        if (amountTotal == 0) revert TokenAmountCannotBeZero();
+        if (endAt != 0 && startAt > endAt) {
+            revert StartCannotBeAfterEnd(startAt, endAt);
+        }
+        if (maxCountPerWallet == 0) revert MaxCountCannotBeZero();
+        IERC1155(releases).safeTransferFrom(_msgSender(), address(this), tokenId, amountTotal, "");
+
+        sales[_msgSender()].push(
+            Sale({
+                releaseOwner: _msgSender(),
+                beneficiary: beneficiary,
+                releases: releases,
+                tokenId: tokenId,
+                amountRemaining: amountTotal,
+                amountTotal: amountTotal,
+                pricePerToken: pricePerToken,
+                startAt: startAt,
+                endAt: endAt,
+                maxCountPerWallet: maxCountPerWallet
+            })
+        );
+
+        emit SaleCreated(_msgSender(), sales[_msgSender()].length - 1);
+    }
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function purchase(
+        address seller,
+        uint256 saleId,
+        uint256 tokenAmount,
+        address recipient
+    ) external nonReentrant {
+        Sale storage sale = _getSaleForPurchase(seller, saleId, tokenAmount);
+
+        uint256 totalPrice = sale.pricePerToken * tokenAmount;
+        uint256 fee = (treasuryFee * totalPrice) / 10_000;
+        IERC20(_usdc).safeTransferFrom(_msgSender(), address(this), totalPrice);
```

Comment:
we shouldn't have to cast _usdc. Let's just make it `token.safeTransferFrom(...)`.